huaweicloud/data_source_huaweicloud_cce_node_pool_v3.go

```diff
@@ -0,0 +1,261 @@
+	config := meta.(*config.Config)
+	cceClient, err := config.CceV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Unable to create HuaweiCloud CCE client : %s", err)
+	}
+
+	listOpts := nodepools.ListOpts{
+		Uid:   d.Get("node_pool_id").(string),
+		Name:  d.Get("name").(string),
+		Phase: d.Get("status").(string),
+	}
+
+	if v, ok := d.GetOk("name"); ok {
+		listOpts.Name = v.(string)
+	}
+
+	if v, ok := d.GetOk("node_pool_id"); ok {
+		listOpts.Uid = v.(string)
+	}
+
+	if v, ok := d.GetOk("status"); ok {
+		listOpts.Phase = v.(string)
+	}
+
+	refinedNodePools, err := nodepools.List(cceClient, d.Get("cluster_id").(string), listOpts)
+
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve Node Pools: %s", err)
+	}
+
+	if len(refinedNodePools) < 1 {
+		return fmt.Errorf("Your query returned no results. " +
+			"Please change your search criteria and try again.")
+	}
+
+	if len(refinedNodePools) > 1 {
+		return fmt.Errorf("Your query returned more than one result." +
+			" Please try a more specific search criteria")
+	}
+
+	NodePool := refinedNodePools[0]
+
+	log.Printf("[DEBUG] Retrieved Node Pools using given filter %s: %+v", NodePool.Metadata.Id, NodePool)
+	d.SetId(NodePool.Metadata.Id)
+	d.Set("node_pool_id", NodePool.Metadata.Id)
+	d.Set("name", NodePool.Metadata.Name)
+	d.Set("type", NodePool.Spec.Type)
+	d.Set("flavor_id", NodePool.Spec.NodeTemplate.Flavor)
+	d.Set("availability_zone", NodePool.Spec.NodeTemplate.Az)
+	d.Set("os", NodePool.Spec.NodeTemplate.Os)
+	d.Set("billing_mode", NodePool.Spec.NodeTemplate.BillingMode)
+	d.Set("key_pair", NodePool.Spec.NodeTemplate.Login.SshKey)
+	d.Set("scall_enable", NodePool.Spec.Autoscaling.Enable)
+	d.Set("initial_node_count", NodePool.Spec.InitialNodeCount)
+	d.Set("current_node_count", NodePool.Status.CurrentNode)
+	d.Set("min_node_count", NodePool.Spec.Autoscaling.MinNodeCount)
+	d.Set("max_node_count", NodePool.Spec.Autoscaling.MaxNodeCount)
+	d.Set("scale_down_cooldown_time", NodePool.Spec.Autoscaling.ScaleDownCooldownTime)
+	d.Set("priority", NodePool.Spec.Autoscaling.Priority)
+	d.Set("subnet_id", NodePool.Spec.NodeTemplate.NodeNicSpec.PrimaryNic.SubnetId)
+	d.Set("max_pods", NodePool.Spec.NodeTemplate.ExtendParam["maxPods"])
+	d.Set("extend_param", NodePool.Spec.NodeTemplate.ExtendParam)
+	d.Set("status", NodePool.Status.Phase)
+	d.Set("region", GetRegion(d, config))
+
+	labels := map[string]string{}
+	for key, val := range NodePool.Spec.NodeTemplate.K8sTags {
+		if strings.Contains(key, "cce.cloud.com") {
+			continue
+		}
+		labels[key] = val
+	}
+	d.Set("labels", labels)
+
+	var volumes []map[string]interface{}
+	for _, pairObject := range NodePool.Spec.NodeTemplate.DataVolumes {
+		volume := make(map[string]interface{})
+		volume["size"] = pairObject.Size
+		volume["volumetype"] = pairObject.VolumeType
+		volume["extend_params"] = pairObject.ExtendParam
+		volumes = append(volumes, volume)
```

Comment:
```
var volumes []map[string]interface{} = make([]map[string]interface{}, len(NodePool.Spec.NodeTemplate.DataVolumes))
```
  or
```
volumes := make([]map[string]interface{}, len(NodePool.Spec.NodeTemplate.DataVolumes))
```
Using append without declaring the size of the memory space in advance will cause memory waste.
Use index such as `volumes [key]` should be better.

Reply:
fixed